contracts/NameWrapper.sol

```diff
@@ -573,47 +577,75 @@ contract NameWrapper is Ownable, ERC1155Fuse, INameWrapper, IERC721Receiver {
     }
 
     /**
-     * @dev Internal function that checks all a name's ancestors to ensure fuse values will be respected.
+     * @dev Internal function that checks all a name's ancestors to ensure fuse values will be respected and parent controller/registrant are set to the Wrapper
      * @param name The name to check.
      * @param offset The offset into the name to start at.
      * @return node The calculated namehash for this part of the name.
-     * @return expiration The timestamp at which the name or its parent expires.
+     * @return vulnerability The timestamp at which the name or its parent expires.
      */
-    function _getExpiration(bytes memory name, uint256 offset)
+    function _checkHierarchy(bytes memory name, uint256 offset)
         internal
         view
-        returns (bytes32 node, uint256 expiration)
+        returns (
+            bytes32 node,
+            ParentVulnerability vulnerability,
+            bytes32 vulnerableNode
+        )
     {
         // Read the first label. If it's the root, return immediately.
         (bytes32 labelhash, uint256 newOffset) = name.readLabel(offset);
         if (labelhash == bytes32(0)) {
             // Root node
-            return (bytes32(0), type(uint256).max);
+            return (bytes32(0), ParentVulnerability.Safe, 0);
         }
 
         // Check the parent name
         bytes32 parentNode;
-        (parentNode, expiration) = _getExpiration(name, newOffset);
+        (
+            parentNode, /* 0 */
+            vulnerability,
+            vulnerableNode
+        ) = _checkHierarchy(name, newOffset);
+
         node = _makeNode(parentNode, labelhash);
 
-        if (expiration < block.timestamp) {
-            // If expiration is less than now, return immediately; no need to check fuses.
-            return (node, expiration);
+        // stop function checking any other nodes if a parent is not safe
+        if (vulnerability != ParentVulnerability.Safe) {
+            return (node, vulnerability, vulnerableNode);
         }
 
         // Check the parent name's fuses to see if replacing subdomains is forbidden
         if (parentNode == ROOT_NODE) {
             // Save ourselves some gas; root node can't be replaced
-            return (node, expiration);
+            return (node, ParentVulnerability.Safe, 0);
         } else if (parentNode == ETH_NODE) {
-            // Special case .eth: Get the expiration from the registrar
-            expiration = registrar.nameExpires(uint256(labelhash));
-            return (node, expiration);
+            // Special case .eth: Check registrant or name isexpired
+
+            try registrar.ownerOf(uint256(labelhash)) returns (
```

Comment:
It might be worth putting both ownership checks in a separate function called from here.

Reply:
Do we want to do that? We still need to check the owner if the name is not .eth, so if we had it in a separate function we couldn't fall through to the `ens.owner` call below

Reply:
I'm not sure I understand. Why can't we just do both checks in that function, depending on whether it's a .eth name or not?